### PR TITLE
Updated java image tests and code to take buffered images instead of lists of lists

### PR DESCRIPTION
**Pull request by anneCarlson**
This will be merged onto annie/text-tags. Take a look when you have a chance, @madisonmay

### CODE REVIEW

**Comment by anneCarlson** [comment]
Note that all the added tests return 400 errors right now. We may want to push the IndicoApi updates to master and make sure these run before merging this anywhere.

**Comment by madisonmay** [comment]
Yup, I agree.

**Comment by madisonmay** [comment]
Should I start code review for this portion so we don't have too giant of a diff all at once?

**Comment by madisonmay** [comment]
We might want to rename the auth keyword to be more general, since I can imagine something like our new `ignore_errors` flag ending up in the same HashMap.

**Comment by anneCarlson** [comment]
I agree. How about `config`?

File: src/main/java/io/indico/api/image/IndicoImageService.java [hunk 1]
```diff
@@ -15,6 +9,7 @@
 
 import java.util.Map;
 import java.util.List;
+import java.awt.image.BufferedImage;
```
Comment (madisonmay) [inline src/main/java/io/indico/api/image/IndicoImageService.java]: I don't think this import is necessary.

**Comment by anneCarlson** [comment]
@madisonmay ready for review before merging. I'm going to update username password to api_key once this is merged one more level in.

**Comment by madisonmay** [comment]
Perfect, I'll give this a look immediately after I locate some lunch.

File: pom.xml [hunk 2]
```diff
@@ -56,6 +56,12 @@
             <scope>test</scope>
         </dependency>
```
Comment (madisonmay) [inline pom.xml]: We should also tick the version number in this file.

File: src/main/java/io/indico/api/image/IndicoImageApi.java [hunk 3]
```diff
@@ -59,223 +58,159 @@ public static IndicoImageApi getInstance() {
...
-
-            if (features == null) {
-                throw ExceptionFactory.getParameterException(facialFeatures.getErrorMessage());
-            }
-
-            return features;
-
-        } catch (RetrofitError error) {
-            throw ExceptionFactory.get(error);
+    private IndicoImageService whichImageService(Map<String, String> auth) {
```
Comment (madisonmay) [inline src/main/java/io/indico/api/image/IndicoImageApi.java]: Why is this function specific to images?

Comment (madisonmay) [inline src/main/java/io/indico/api/image/IndicoImageApi.java]: `auth.get("cloud")` should return a subdomain name not a full url -- we need to then do string formatting to send requests to sudomain.indico.domains.

Comment (madisonmay) [inline src/main/java/io/indico/api/image/IndicoImageApi.java]: Wait nevermind I think I get what's going on now --- I see the logic for that in `IndicoImageApi` now.

File: src/main/java/io/indico/api/text/IndicoTextApi.java [hunk 4]
```diff
@@ -59,39 +58,16 @@ public static IndicoTextApi getInstance() {
...
-     */
-    public Map<String, Double> language(String text) throws IndicoException {
-        try {
-            return this.indicoTextService.language(new Text(text)).get("results");
-        } catch (RetrofitError error) {
-            throw ExceptionFactory.get(error);
+     * Returns either an indicoTextService with the endpoint
+     * or, if it exists, the endpoint set in the parameter 'cloud' in auth
+     */
+    private IndicoTextService whichTextService(Map<String, String> auth) {
```
Comment (madisonmay) [inline src/main/java/io/indico/api/text/IndicoTextApi.java]: Feels like we're repeating logic in `IndicoImageApi` here.

File: src/main/java/io/indico/api/text/IndicoTextApi.java [hunk 5]
```diff
@@ -638,21 +484,33 @@ public void failure(RetrofitError retrofitError) {
...
+            return thisIndicoTextService.batchSentiment(new TextList(text), encodeCredentials(auth.get("username"), auth.get("password"))).get("results");
         } catch (RetrofitError error) {
             throw ExceptionFactory.get(error);
         }
     }
 
-    private static String encodeCredentials(String username, String password) {
+
+    private static String encodeCredentials(String username, String password) throws IndicoException {
+        if (username == null) {
```
Comment (madisonmay) [inline src/main/java/io/indico/api/text/IndicoTextApi.java]: A bit more repeated code (although this is going to disappear soon, so I wouldn't worry too much about it).

File: src/main/resources/data/testIniComplete [hunk 6]
```diff
@@ -0,0 +1,6 @@
+[private_cloud]
```
Comment (madisonmay) [inline src/main/resources/data/testIniComplete]: These will also have to change for the move to API keys.

File: pom.xml (hunk 2 above)
Comment (anneCarlson) [inline pom.xml]: Good point. I'll get on that.

File: src/test/java/io/indico/api/test/image/IndicoImageImageApiTest.java [hunk 7]
```diff
@@ -1,66 +1,64 @@
...
 
-import java.util.*;
+public class IndicoImageImageApiTest {
 
-import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertTrue;
+    private BufferedImage img;
+
+    @Before
+    public void loadImage() {
```
Comment (madisonmay) [inline src/test/java/io/indico/api/test/image/IndicoImageImageApiTest.java]: Super nitpicky but we could also pull this bit of code out so we don't have a half dozen copies.

File: src/test/java/io/indico/api/test/unit/configTest.java [hunk 8]
```diff
@@ -0,0 +1,147 @@
+package io.indico.api.test.unit;
```
Comment (madisonmay) [inline src/test/java/io/indico/api/test/unit/configTest.java]: Awesome suite of tests!

File: src/main/java/io/indico/api/image/IndicoImageApi.java (hunk 3 above)
Comment (anneCarlson) [inline src/main/java/io/indico/api/image/IndicoImageApi.java]: Right now we have IndicoTextService and IndicoImageService separate. There is no reason they have to be; I think they were just set up that way to keep files a reasonable length and that seemed like a reasonable division. whichImageService returns an ImageService and whichTextService returns a TextService

File: src/main/java/io/indico/api/text/IndicoTextApi.java (hunk 4 above)
Comment (anneCarlson) [inline src/main/java/io/indico/api/text/IndicoTextApi.java]: We are, unfortunately, as Java methods can only return one type, the way things are currently written these need to be separate methods. We could make IndicoTextService and IndicoImageService inherent from the same interface somehow (I'm not certain how interface inheritance works in Java but I could look into it).

File: src/main/java/io/indico/api/text/IndicoTextApi.java (hunk 5 above)
Comment (anneCarlson) [inline src/main/java/io/indico/api/text/IndicoTextApi.java]: Yeah, when we change to apikeys I'll put this method out to somewhere I could import into both these classes.

File: src/test/java/io/indico/api/test/image/IndicoImageImageApiTest.java (hunk 7 above)
Comment (anneCarlson) [inline src/test/java/io/indico/api/test/image/IndicoImageImageApiTest.java]: Agreed. I'll pull that out. That could get really annoying as we add more tests or change something about this bit of code.

**Comment by anneCarlson** [comment]
@madisonmay Made the updates we talked about. Just give me the go ahead and I'll merge.

**Comment by anneCarlson** [comment]
I'm actually just going to merge this. The changes were minor and I'm just merging into another branch, not master.

**Comment by madisonmay** [comment]
Go for it!  Sorry for the slow response time.

**Pull request merged**